books/RayTracingTheNextWeek.html

```diff
@@ -2256,7 +2256,319 @@
+            D = -dot(normal, corner);
+            w = n / dot(n,n);
+
+            set_bounding_box();
+        }
+
+        virtual void set_bounding_box() {
+            bbox = aabb(corner, corner + side_A + side_B).pad();
+        }
+
+        aabb bounding_box() const override { return bbox; }
+
+        virtual bool hit_ab(double a, double b, hit_record& rec) const {
+            // Given the hit point in plane coordinates, return false if it is outside the
+            // primitive, otherwise set the hit record UV coordinates and return true.
+
+            if ((a < 0) || (1 < a) || (b < 0) || (1 < b))
+                return false;
+
+            rec.u = a;
+            rec.v = b;
+            return true;
+        }
+
+        bool hit(const ray& r, interval ray_t, hit_record& rec) const override {
+            auto denom = dot(normal, r.direction());
+
+            // No hit if the ray is parallel to the plane.
+            if (fabs(denom) < 1e-8)
+                return false;
+
+            // Return false if the hit point parameter t is outside the ray interval.
+            auto t = (-D - dot(normal, r.origin())) / denom;
+            if (!ray_t.contains(t))
+                return false;
+
+            // Determine the hit point lies within the planar shape using its plane coordinates.
+            auto intersection = r.at(t);
+            vec3 planar_hitpt_vector = intersection - corner;
+            auto a = dot(w, cross(planar_hitpt_vector, side_B));
+            auto b = dot(w, cross(side_A, planar_hitpt_vector));
+
+            if (!hit_ab(a, b, rec))
+                return false;
+
+            // Ray hits the 2D shape; set the rest of the hit record and return true.
+            rec.t = t;
+            rec.p = intersection;
+            rec.mat = mat;
+            rec.set_face_normal(r, normal);
+
+            return true;
+        }
+
+      protected:
+        point3 corner;
+        vec3 side_A, side_B;
+        shared_ptr<material> mat;
+        vec3 normal;
+        double D;
+        vec3 w;
+        aabb bbox;
+    };
+
+    #endif
+    ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+    [Listing [quad]: <kbd>[quad.h]</kbd> 2D Quadrilateral (Parallelogram) Primitives]
+</div>
+
+
+More Planar Primitives
+-----------------------
+You may have noticed that if you have the 2D planar coordinates of a hit point, you can implement
+arbitrary 2D shapes -- not just parallelograms. In fact, you'd only need to change the `hit_ab`
+method above to determine if the (a,b) coordinates lie inside your 2D primitive, and set the texture
+UV coordinates if so.
+
+We'll leave these additional 2D primitives as an exercise to the reader, depending on your desire to
+explore. Consider triangles, disks, and rings. You could even create cut-out stencils based on the
+texture map! We've included the solution for triangle, ellipse, and annulus in the commit history
```

Comment:
We've included the solution for triangle, ellipse, and annulus ...

Consider cutting. The book != the commit history

Reply:
Ok. I'll leave it as an Easter egg.

Reply:
Cut.